balancer/rls/internal/child_policy.go

```diff
@@ -0,0 +1,216 @@
+)
+
+// childPolicyWrapper is a reference counted wrapper around a child policy.
+//
+// The LB policy maintains a map of these wrappers keyed by the target returned
+// by RLS. When a target is seen for the first time, a child policy wrapper is
+// created for it and the wrapper is added to the child policy map. Each entry
+// in the data cache holds references to the corresponding child policy
+// wrappers. The LB policy also holds a reference to the child policy wrapper
+// for the default target specified in the LB Policy Configuration
+//
+// When a cache entry is evicted, it releases references to the child policy
+// wrappers that it contains. When all references have been released, the
+// wrapper is removed from the child policy map and is destroyed.
+//
+// The child policy wrapper also caches the connectivity state and most recent
+// picker from the child policy. Once the child policy wrapper reports
+// TRANSIENT_FAILURE, it will continue reporting that state until it goes READY;
+// transitions from TRANSIENT_FAILURE to CONNECTING are be ignored.
+//
+// Whenever a child policy wrapper changes its connectivity state, the LB policy
+// returns a new picker to the channel, since the channel may need to re-process
+// the picks for queued RPCs.
+//
+// It is not safe for concurrent access.
+type childPolicyWrapper struct {
+	bg     balancerGroup // BalancerGroup to which this child policy belongs.
+	logger *internalgrpclog.PrefixLogger
+
+	target        string                     // RLS target corresponding to this child policy.
+	targetField   string                     // Child policy config field to plug in the above target.
+	config        map[string]json.RawMessage // Child policy config.
+	builder       balancer.Builder           // Child policy builder to be passed to the balancer group.
+	parser        balancer.ConfigParser      // Child policy config parser for building the config.
+	resolverState resolver.State             // Resolver state as reported to the RLS LB policy.
+
+	refCnt           int                // Reference count.
+	state            balancer.State     // Balancer state reported by the child policy.
+	stateToAggregate connectivity.State // Child policy state reported for aggregation.
+}
+
+// childPolicyWrapperArgs is simply a collection of the arguments to be passed
+// to construct or update a child policy wrapper.
+type childPolicyWrapperArgs struct {
+	policyName    string
+	target        string
+	targetField   string
+	config        map[string]json.RawMessage
+	resolverState resolver.State
+	bg            balancerGroup
+}
+
+// newChildPolicyWrapper creates a new child policy wrapper for the given
+// arguments. The wrapper is initialized with no references and starts off in
+// CONNECTING state.
+//
+// It is important to note that the child policy is not added to the balancer
+// group at this point. It is added only when the reference to the child policy
+// is taken.
```

Comment:
Why not add it now?  It seems like that would make life much simpler?

Reply:
Done. Thanks.